File: Dynamic Programming/DP on MCM - Partition DP/Matrix Chain Multiplication/recursive.cpp
int solve(int i, int j, vector<int> &arr) {
    if(i >= j) return 0;
    int mini = 1e9;
    for(int k=i; k<=j-1; k++){
        int temp = arr[i-1]*arr[k]*arr[j] + solve(i, k, arr) + solve(k+1, j, arr);
        mini = min(mini, temp);
    }
    return mini;
}

int matrixMultiplication(vector<int> &arr) {
    int n = arr.size();
    return solve(1, n-1, arr, dp);
}

File: Dynamic Programming/DP on MCM - Partition DP/Matrix Chain Multiplication/recursive.java

class Solution {
    static int matrixMultiplication(int arr[]) {
        return helper(1, arr.length-1, arr);
    }
    
    private static int helper(int i, int j, int[] arr){
        if (i >= j) return 0;
        
        int res = Integer.MAX_VALUE;
        
        for (int k=i; k<j; k++){
            int left = helper(i, k, arr);
            int right = helper(k+1, j, arr);
            int temp = arr[i-1] * arr[k] * arr[j] + left + right;
            
            res = Math.min(res, temp);
        }
        
        return res;
    }
}